enginetest/enginetests.go

```diff
@@ -5459,3 +5461,144 @@ func widenJSONArray(narrow []interface{}) (wide []interface{}) {
 	}
 	return
 }
+
+var (
+	jsonData []byte
+)
+
+// privDataJson is used to marshal/unmarshal the privilege data to/from JSON.
+type privDataJson struct {
+	Users []*grant_tables.User
+	Roles []*grant_tables.RoleEdge
+}
+
+func temporaryLoadPrivileges() ([]*grant_tables.User, []*grant_tables.RoleEdge, error) {
+	if len(jsonData) == 0 {
+		return nil, nil, nil
+	}
+	data := &privDataJson{}
+	err := json.Unmarshal(jsonData, data)
+	if err != nil {
+		return nil, nil, err
+	}
+	return data.Users, data.Roles, nil
+}
+
+var _ grant_tables.PersistCallback = temporarySavePrivileges
+
+func temporarySavePrivileges(ctx *sql.Context, users []*grant_tables.User, roles []*grant_tables.RoleEdge) error {
+	data := &privDataJson{
+		Users: users,
+		Roles: roles,
+	}
+	jd, err := json.Marshal(data)
+	if err != nil {
+		return err
+	}
+	jsonData = jd
+	return nil
+}
+
+func TestPersistPrivileges(t *testing.T, h Harness) {
+	harness, ok := h.(ClientHarness)
+	if !ok {
+		t.Skip("Cannot run TestUserPrivileges as the harness must implement ClientHarness")
+	}
+
+	port := getEmptyPort(t)
+	ctx := NewContextWithClient(harness, sql.Client{
+		User:    "root",
+		Address: "localhost",
+	})
+	serverConfig := server.Config{
+		Protocol:       "tcp",
+		Address:        fmt.Sprintf("localhost:%d", port),
+		MaxConnections: 1000,
+	}
+
+	users, roles, _ := temporaryLoadPrivileges()
+
+	myDb := harness.NewDatabase("mydb")
+	databases := []sql.Database{myDb}
+	engine := NewEngineWithDbs(t, harness, databases)
+	defer engine.Close()
+	engine.Analyzer.Catalog.GrantTables.AddRootAccount()
+	engine.Analyzer.Catalog.GrantTables.SetPersistCallback(temporarySavePrivileges)
+	engine.Analyzer.Catalog.GrantTables.LoadData(ctx, users, roles)
+
+	for _, script := range PersistPrivilegesTests {
+		// start test with empty `privileges file`
+		jsonData = []byte{}
+
+		t.Run(script.Name, func(t *testing.T) {
+			for _, statement := range script.SetUpScript {
+				if sh, ok := harness.(SkippingHarness); ok {
+					if sh.SkipQueryTest(statement) {
+						t.Skip()
+					}
+				}
+				RunQueryWithContext(t, engine, ctx, statement)
+			}
+
+			for _, assertion := range script.InOneServer {
+				runTestsInOneServer(t, harness, ctx, serverConfig, assertion)
+			}
+		})
+	}
+}
+
+func runTestsInOneServer(t *testing.T, harness ClientHarness, ctx *sql.Context, sc server.Config, script PersistPrivilegesTestInOneServer) {
+	users, roles, _ := temporaryLoadPrivileges()
+	myDb := harness.NewDatabase("mydb")
+	databases := []sql.Database{myDb}
+	engine := NewEngineWithDbs(t, harness, databases)
+	defer engine.Close()
+	engine.Analyzer.Catalog.GrantTables.AddRootAccount()
+	engine.Analyzer.Catalog.GrantTables.SetPersistCallback(temporarySavePrivileges)
+	engine.Analyzer.Catalog.GrantTables.LoadData(ctx, users, roles)
+
+	s, err := server.NewDefaultServer(sc, engine)
```

Comment:
I mentioned this in my last feedback, and I see the same issue. This server is started, but it is not being used at all. Not only that, it wouldn't need to be used as this test does not test a workflow that's specific to client connections.

It's fine to model similar tests to pre-existing tests when some of the boilerplate is the same, but you need to make sure that every function call and variable declaration is relevant to _your_ test. For any functions that seem required, if you're not exactly sure how they operate, it's best to dig into them and try and get an understanding. Of course you can always ask for clarification as well, and don't hesitate to ask questions.